compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -292,18 +292,23 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   private val nullType             = ConstantType(Constant(null))
   private val nullSpace            = Typ(nullType)
 
-  override def intersectUnrelatedAtomicTypes(tp1: Type, tp2: Type): Space = {
-    val and = AndType(tp1, tp2)
-    // Precondition: !(tp1 <:< tp2) && !(tp2 <:< tp1)
-    // Then, no leaf of the and-type tree `and` is a subtype of `and`.
-    val res = inhabited(and)
-
-    debug.println(s"atomic intersection: ${and.show} = ${res}")
-
-    if (!res) Empty
-    else if (tp1.isSingleton) Typ(tp1, true)
-    else if (tp2.isSingleton) Typ(tp2, true)
-    else Typ(and, true)
+  override def intersectUnrelatedAtomicTypes(tp1: Type, tp2: Type) = {
+    // Precondition: !isSubType(tp1, tp2) && !isSubType(tp2, tp1)
+    if (tp1 == nullType || tp2 == nullType) {
+      // Since projections of types don't include null, intersection with null is empty.
+      Empty
```

Comment:
A minor comment: it seems more readable if we just `return Empty` for the exceptional case, fewer braces and indentations.